### ISSUE TITLE
[Build] Console output lost

### ISSUE THREAD

**Issue by ManickaP** [dotnet/runtime#97211]
After update to the newest tools, the output from msbuild has changed and now auto-updates an existing line in terminal instead of printing all output out. As a result, the important information about failing tests is lost and this is what remains:
```sh
dotnet build tests/FunctionalTests/ -t:test
MSBuild version 17.10.0-preview-24060-03+1725b247e for .NET
Restore complete (2.6s)
  TestUtilities net8.0 succeeded (4.2s) → /home/manicka/repositories/runtime/artifacts/bin/TestUtilities/Debug/net8.0/TestUtilities.dll
  System.Net.Quic.Functional.Tests net9.0-linux failed with errors (30.3s) → /home/manicka/repositories/runtime/artifacts/bin/System.Net.Quic.Functional.Tests/Debug/net9.0-linux/System.Net.Quic.Functional.Tests.dll
    /home/manicka/repositories/runtime/eng/testing/tests.targets(195,5): error : One or more tests failed while running tests from 'System.Net.Quic.Functional.Tests'. [/home/manicka/repositories/runtime/src/libraries/System.Net.Quic/tests/FunctionalTests/System.Net.Quic.Functional.Tests.csproj]

Build failed with errors in 40.1s
```

No failing test name, no failure details, no clickable link to `testResults.xml`.

I do not know if this is an issue in our infra (the way we call msbuild) or arcade or msbuild itself, so feel free to transfer this issue.

**Comment by akoeplinger** [comment]
That's the new msbuild terminal logger (https://github.com/dotnet/sdk/pull/37492), you can disable it via `-tl:off`.

Looks like there are some improvements coming for dotnet test as well: https://github.com/dotnet/sdk/pull/38098

**Comment by stephentoub** [comment]
> Maybe we should disable the terminal logger for the test target for now.

Yes, please. I just got bit by this as well.

Why is not showing the test results a better experience, @rainersigwald?

**Comment by rainersigwald** [comment]
## Local workaround

Set the environment variable `DOTNET_CLI_CONFIGURE_MSBUILD_TERMINAL_LOGGER=false` which will opt out of the new logger.

## Committable workaround

> Maybe we should disable the terminal logger for the test target for now.

You can't do this based on requested target; you'll need to disable -tl entirely (easiest way would be to put `-tl:false` in a root `Directory.Build.rsp`).

## Ok but what's going on?

> I'm not sure though if that will automatically apply to our infrastructure.

They would not, the critical change there is what logging/warning/error APIs the task running the tests calls, and Arcade tests do that all manually.

> Why is not showing the test failures a better experience, @rainersigwald?

It isn't. What happened is that build output is so polluted by "high" priority messages that the terminal logger mostly disregards them. Here, it sounds like all of the debugging test output is given in a message, rather than in the error, which combines badly with that.

The plan for `dotnet test` is that the test runner task will switch to using MSBuild logging rather than directly emitting to stdout, and we're planning to add logging APIs to allow test to produce a nicely formatted summary message (like they do today by bypassing MSBuild, but not lost to logs and multiproc-compatible).

It should be possible to fix the repo/Arcade test infrastructure to provide more details in a way that is more MSBuild-y. I'll take a look.

cc @baronfel, @rokonec

**Comment by rainersigwald** [comment]
There's already logic that tries to include the TRX in the failure message, which works for me when I artificially induce a failure in STJ tests.

```sh-session
❯ dotnet build -t:test -bl
MSBuild version 17.10.0-preview-24060-03+1725b247e for .NET
Restore complete (1.0s)
...
  System.Text.Json net462 succeeded (0.7s) → S:\runtime\artifacts\bin\System.Text.Json\Debug\net462\System.Text.Json.dll
  System.Text.Json.Tests net462 failed with errors (1.4s) → S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\System.Text.Json.Tests.dll
    S:\runtime\eng\testing\tests.targets(195,5): error : One or more tests failed while running tests from 'System.Text.Json.Tests'. Please check S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\testResults.xml for details! [S:\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\System.Text.Json.Tests.csproj]

Build failed with errors in 8.4s
```

I don't know why that didn't trigger for you, @ManickaP.

Of course, that's still less data than the super verbose dump you had before

```
  System.Text.Json -> S:\runtime\artifacts\bin\System.Text.Json\Debug\net462\System.Text.Json.dll
  System.Text.Json.Tests -> S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\System.Text.Json.Tests.dll
  ========================= Begin custom configuration settings ==============================
  ========================== End custom configuration settings ===============================
  ----- start Fri 01/19/2024 11:46:27.91 ===============  To repro directly: ==========================================
  ===========
  pushd S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\
  xunit.console.exe System.Text.Json.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait catego
  ry=failing
  popd
  ===========================================================================================================
    Discovering: System.Text.Json.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method disp
  lay options = None)
    Discovered:  System.Text.Json.Tests (found 9 test cases)
    Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 32)
      System.Text.Json.Tests.JsonElementCloneTests.CloneTwiceFromSameDocument [FAIL]
        Assert.True() Failure
        Expected: True
        Actual:   False
        Stack Trace:
          S:\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\JsonElementCloneTests.cs(14,0): at Sys
  tem.Text.Json.Tests.JsonElementCloneTests.CloneTwiceFromSameDocument()
    Finished:    System.Text.Json.Tests
  === TEST EXECUTION SUMMARY ===
     System.Text.Json.Tests  Total: 9, Errors: 0, Failed: 1, Skipped: 0, Time: 0.180s
  ----- end Fri 01/19/2024 11:46:28.41 ----- exit code 1 ----------------------------------------------------------
S:\runtime\eng\testing\tests.targets(195,5): error : One or more tests failed while running tests from 'System.Text.Jso
n.Tests'. Please check S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\testResults.xml for details! [S:\ru
ntime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\System.Text.Json.Tests.csproj::TargetFramework=net462
]

Build FAILED.
```

Trying to see if we can conditionally emit that on test failure.

**Comment by stephentoub** [comment]
> I don't know why that didn't trigger for you

It says that there are errors, but not what those errors are. That's my problem. To actually see the errors I then need to go load up the XML file from xunit and search through it looking for what went wrong.

**Comment by stephentoub** [comment]
> I'll take a look.

Thanks!

**Comment by rainersigwald** [comment]
How would y'all feel about this?

![image](https://github.com/dotnet/runtime/assets/3347530/ca3751cb-73c3-4087-9fc4-904f59b96c15)

<details>
<summary>text version</summary>

```
  System.Text.Json net462 succeeded (0.1s) → S:\runtime\artifacts\bin\System.Text.Json\ref\Debug\net462\System.Text.Json.dll
  System.Text.Json net462 succeeded (0.9s) → S:\runtime\artifacts\bin\System.Text.Json\Debug\net462\System.Text.Json.dll
--interactive test output, because a test failed--
========================= Begin custom configuration settings ==============================
========================== End custom configuration settings ===============================
----- start Fri 01/19/2024 13:44:41.68 ===============  To repro directly: =====================================================
pushd S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\
xunit.console.exe System.Text.Json.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
popd
===========================================================================================================
Discovering: System.Text.Json.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method display options = None)
Discovered:  System.Text.Json.Tests (found 9 test cases)
Starting:    System.Text.Json.Tests (parallel test collections = on, max threads = 32)
System.Text.Json.Tests.JsonElementCloneTests.CloneTwiceFromSameDocument [FAIL]
Assert.True() Failure
Expected: True
Actual:   False
Stack Trace:
S:\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\JsonElementCloneTests.cs(14,0): at System.Text.Json.Tests.JsonElementCloneTests.CloneTwiceFromSameDocument()
Finished:    System.Text.Json.Tests
=== TEST EXECUTION SUMMARY ===
System.Text.Json.Tests  Total: 9, Errors: 0, Failed: 1, Skipped: 0, Time: 0.188s
----- end Fri 01/19/2024 13:44:42.26 ----- exit code 1 ----------------------------------------------------------
  System.Text.Json.Tests net462 failed with errors (1.9s) → S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\System.Text.Json.Tests.dll
    S:\runtime\eng\testing\tests.targets(199,5): error : One or more tests failed while running tests from 'System.Text.Json.Tests'. Please check S:\runtime\artifacts\bin\System.Text.Json.Tests\Debug\net462\testResults.xml for details! [S:\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\System.Text.Json.Tests.csproj]

Build failed with errors in 17.4s
```

</details>

It loses the leading whitespace (https://github.com/dotnet/msbuild/issues/9668) but is fairly close to the prior output and can be done with a small patch in this repo (d1f74f4a6c3).

**Comment by wfurt** [comment]
> It says that there are errors, but not what those errors are. That's my problem. To actually see the errors I then need to go load up the XML file from xunit and search through it looking for what went wrong.

note that CI test runs do not preserver the XML any more AFAIK. e.g. the console output is only once clue we have in many cases.

**Comment by wfurt** [comment]
> How would y'all feel about this?

do you have example with deeper stack @rainersigwald ?

**Comment by rainersigwald** [comment]
> the console output is only once clue we have in many cases.

The terminal logger should not be involved in CI test runs--we should detect that output is being redirected and fall back to the older output behavior. If you see otherwise that is a bug in MSBuild.

But this is good to know--does it mean you would want this output **all the time**, not just on test failure?

**Comment by rainersigwald** [comment]
> > How would y'all feel about this?
> 
> do you have example with deeper stack @rainersigwald ?

Sure:

![image](https://github.com/dotnet/runtime/assets/3347530/fb0d2d26-f0ac-4488-9ece-d6e63acbdfe2)

**Comment by stephentoub** [comment]
> How would y'all feel about this?

Not ideal, since without color the whitespace is the main indicator of where one failure ends and the next begins. It's a lot harder to skim the output for pertinent info. But certainly better than no test output at all.

**Comment by Joe4evr** [comment]
> > the console output is only once clue we have in many cases.
> 
> The terminal logger should not be involved in CI test runs--we should detect that output is being redirected and fall back to the older output behavior. If you see otherwise that is a bug in MSBuild.

For reference, a code base at my work has been struggling with the lack of test run failure information in CI for like half a year, and that has been building against .NET 6. I haven't been able to find a single log file or anything that even points to *why* the CI `dotnet` process fails out of (seemingly) nowhere, and yet when running tests locally they all finish which makes it even more impossible to diagnose.

**Comment by stephentoub** [comment]
> My failure was a crash (`Debug.Assert` not test assert). But I learned that later after filing this. I haven't had any details in the output and with crash there's no `testResults.xml`.

@akoeplinger, could you help disable the new thing repo wide until these issues are addressed? Thanks!

@rainersigwald, let's make sure crashes/asserts are visible as well. Thanks in advance for fixing it!

**Comment by AR-May** [comment]
We plan to make terminal logger to respect verbosity level. After this feature is done, high-importance messages would be accessible if one set verbosity to detailed or diagnostic (default dotnet verbosity setting is `minimal`). 

Here is the example of the more verbose run with the terminal logger: 
```
PS C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\src\libraries\System.Collections.Immutable\tests> C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\.dotnet\dotnet.exe build /t:Test -v:d
MSBuild version 17.10.0-dev-24163-01+582d6db85 for .NET
Restore complete (3.6s)
    Determining projects to restore...
    All projects are up-to-date for restore.
  TestUtilities net8.0 succeeded (0.5s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\TestUtilities\Debug\net8.0\TestUtilities.dll
  ILLink.RoslynAnalyzer succeeded (0.2s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\ILLink.RoslynAnalyzer\Debug\netstandard2.0\ILLink.RoslynAnalyzer.dll
  ILLink.CodeFixProvider succeeded (0.2s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\ILLink.CodeFixProvider\Debug\netstandard2.0\ILLink.CodeFixProvider.dll
  Mono.Linker succeeded (0.2s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\Mono.Linker\ref\Debug\net9.0\illink.dll
  Mono.Linker succeeded (0.3s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\Mono.Linker\Debug\net9.0\illink.dll
  ILLink.Tasks net9.0 succeeded (0.2s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\ILLink.Tasks\Debug\net9.0\ILLink.Tasks.dll
  Microsoft.Interop.SourceGeneration succeeded (0.3s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\Microsoft.Interop.SourceGeneration\Debug\netstandard2.0\Microsoft.Interop.SourceGeneration.dll
  LibraryImportGenerator succeeded (0.3s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\LibraryImportGenerator\Debug\netstandard2.0\Microsoft.Interop.LibraryImportGenerator.dll
  ComInterfaceGenerator succeeded (0.3s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\ComInterfaceGenerator\Debug\netstandard2.0\Microsoft.Interop.ComInterfaceGenerator.dll
  System.Runtime succeeded (0.1s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Runtime\ref\Debug\net9.0\System.Runtime.dll
  System.Collections succeeded (0.1s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections\ref\Debug\net9.0\System.Collections.dll
  System.Collections.Immutable net9.0 succeeded (0.1s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable\ref\Debug\net9.0\System.Collections.Immutable.dll
  System.Collections.Immutable net9.0 succeeded (0.2s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable\Debug\net9.0\System.Collections.Immutable.dll
  System.Collections.Immutable.Tests net9.0 succeeded (377.5s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net9.0\System.Collections.Immutable.Tests.dll
    "C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net9.0\RunTests.cmd" --runtime-path "C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\testhost\net9.0-windows-Debug-x64"
    ========================= Begin custom configuration settings ==============================
    set __IsXUnitLogCheckerSupported=1
    ========================== End custom configuration settings ===============================
    ----- start Thu 03/14/2024 13:14:32.99 ===============  To repro directly: =====================================================
    pushd C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net9.0\
    "C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\testhost\net9.0-windows-Debug-x64\dotnet.exe" exec --runtimeconfig System.Collections.Immutable.Tests.runtimeconfig.json --depsfile System.Collections.Immutable.Tests.deps.json xunit.console.dll System.Collections.Immutable.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
    popd
    ===========================================================================================================
      Discovering: System.Collections.Immutable.Tests (method display = ClassAndMethod, method display options = None)
      Discovered:  System.Collections.Immutable.Tests (found 4279 of 4384 test cases)
      Starting:    System.Collections.Immutable.Tests (parallel test collections = on [16 threads], stop on fail = off)
      Finished:    System.Collections.Immutable.Tests
    === TEST EXECUTION SUMMARY ===
       System.Collections.Immutable.Tests  Total: 20564, Errors: 0, Failed: 0, Skipped: 0, Time: 252.670s
    ----- end Thu 03/14/2024 13:20:49.77 ----- exit code 0 ----------------------------------------------------------
  TestUtilities net462 succeeded (0.9s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\TestUtilities\Debug\net462\TestUtilities.dll
  System.Collections.Immutable net462 succeeded (0.1s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable\ref\Debug\net462\System.Collections.Immutable.dll
  System.Collections.Immutable net462 succeeded (0.1s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable\Debug\net462\System.Collections.Immutable.dll
  System.Collections.Immutable.Tests net462 succeeded (57.6s) → C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net462\System.Collections.Immutable.Tests.dll
    "C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net462\RunTests.cmd"
    ========================= Begin custom configuration settings ==============================
    ========================== End custom configuration settings ===============================
    ----- start Thu 03/14/2024 13:20:51.62 ===============  To repro directly: =====================================================
    pushd C:\Users\alinama\work\msbuild\test_repos\runtime\runtime\artifacts\bin\System.Collections.Immutable.Tests\Debug\net462\
    xunit.console.exe System.Collections.Immutable.Tests.dll -xml testResults.xml -nologo -notrait category=OuterLoop -notrait category=failing
    popd
    ===========================================================================================================
      Discovering: System.Collections.Immutable.Tests (app domain = on [no shadow copy], method display = ClassAndMethod, method display options = None)
      Discovered:  System.Collections.Immutable.Tests (found 4009 of 4124 test cases)
      Starting:    System.Collections.Immutable.Tests (parallel test collections = on [16 threads], stop on fail = off)
      Finished:    System.Collections.Immutable.Tests
    === TEST EXECUTION SUMMARY ===
       System.Collections.Immutable.Tests  Total: 20246, Errors: 0, Failed: 0, Skipped: 0, Time: 7.304s
    ----- end Thu 03/14/2024 13:21:48.75 ----- exit code 0 ----------------------------------------------------------

Build succeeded in 446.6s
```

Currently the verbosity level is set to `minimal` in this repository, as I can see, so to use this feature you will need to set it to `detailed`. (This will result in larger logs though in the pipeline when the TL gets automatically disabled because of the redirection of the output). Would that be an acceptable permanent fix instead of opting out of the terminal logger completely?

**Comment by ManickaP** [comment]
This is more question for our infra @akoeplinger @ViktorHofer.

**Comment by akoeplinger** [comment]
We definitely can't switch to `detailed` as the default for the whole build. It might be acceptable if we could set it for a specific target.

**Comment by AR-May** [comment]
I talked to @rainersigwald and we have got one more idea here: we will allow to set a specific verbosity for the Terminal Logger with `/tl:verbosity`. In this case if the TL is disabled, due to redirection or other reasons, the console logger will be called with the verbosity you have set with `/v` for the build, which is `minimal`. And if the TL is used it will be called with detailed verbosity, which will display for you high-prio messages, warnings and errors, which is almost equivalent to the output of Console Logger with minimal verbosity. So, in all cases you will get a reasonable output. Will this work for you, @akoeplinger?
